pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -195,13 +195,17 @@ func (p *staticPolicy) validateState(s state.State) error {
 	tmpAssignments := s.GetCPUAssignments()
 	tmpDefaultCPUset := s.GetDefaultCPUSet()
 
+	allCPUs := p.topology.CPUDetails.CPUs()
+	if p.options.StrictCPUReservation {
+		allCPUs = allCPUs.Difference(p.reservedCPUs)
+	}
+
 	// Default cpuset cannot be empty when assignments exist
 	if tmpDefaultCPUset.IsEmpty() {
 		if len(tmpAssignments) != 0 {
 			return fmt.Errorf("default cpuset cannot be empty")
 		}
 		// state is empty initialize
-		allCPUs := p.topology.CPUDetails.CPUs()
 		s.SetDefaultCPUSet(allCPUs)
 		return nil
 	}
```

Comment:
this is a good chance to move the initialization outside the validation. The validation code should really only read data from the the state, not set it through. I think it boils down to move these lines in the (one and only) call site of `validateState`. We can even extract a new helper if this makes testing easier (it can)

Reply:
/cc @rphillips @haircommander (FYI)

Reply:
I'm thinking about something like
```
$ git diff
diff --git a/pkg/kubelet/cm/cpumanager/policy_static.go b/pkg/kubelet/cm/cpumanager/policy_static.go
index 19c98af703e..27da597b36f 100644
--- a/pkg/kubelet/cm/cpumanager/policy_static.go
+++ b/pkg/kubelet/cm/cpumanager/policy_static.go
@@ -184,6 +184,7 @@ func (p *staticPolicy) Name() string {
 }
 
 func (p *staticPolicy) Start(s state.State) error {
+       p.initialize(s)
        if err := p.validateState(s); err != nil {
                klog.ErrorS(err, "Static policy invalid state, please drain node and remove policy state file")
                return err
@@ -191,6 +192,15 @@ func (p *staticPolicy) Start(s state.State) error {
        return nil
 }
 
+func (p *staticPolicy) initialize(s state.State) {
+       if !s.GetDefaultCPUSet().IsEmpty() {
+               return
+       }
+       cpus := p.topology.CPUDetails.CPUs()
+       s.SetDefaultCPUSet(cpus)
+       klog.InfoS("Static policy initialized", "defaultCPUSet", cpus)
+}
+
 func (p *staticPolicy) validateState(s state.State) error {
        tmpAssignments := s.GetCPUAssignments()
        tmpDefaultCPUset := s.GetDefaultCPUSet()
@@ -200,9 +210,6 @@ func (p *staticPolicy) validateState(s state.State) error {
                if len(tmpAssignments) != 0 {
                        return fmt.Errorf("default cpuset cannot be empty")
                }
-               // state is empty initialize
-               allCPUs := p.topology.CPUDetails.CPUs()
-               s.SetDefaultCPUSet(allCPUs)
                return nil
        }
 ```
this will require first a review of the unit tests of the `Start` method, and adding quite some tests, likely.
However this is a good thing in general and also something we want already for this PR.

Reply:
Hi @ffromani, thank you for the review. What you ask is code refactoring for better housekeeping, not tied to the feature content itself. We will take a look, but we are not sure we are capable of managing this given we are new to CPUManager UT and are adding our first few UT test cases in 2 weeks to meet the release code freeze date.